controllers/ocsinitialization/ocsoperator_config.go

```diff
@@ -0,0 +1,147 @@
+)
+
+const (
+	// This configmap is purely for the OCS operator to use
+	ocsOperatorConfigName = "ocs-operator-config"
+)
+
+// ensureOCSOperatorConfig ensures that the OCS operator config is present in the
+// cluster & has the correct values
+func (r *OCSInitializationReconciler) ensureOCSOperatorConfig(initialData *ocsv1.OCSInitialization) error {
+	ocsOperatorConfig := &corev1.ConfigMap{}
+	err := r.Client.Get(context.TODO(), types.NamespacedName{Name: ocsOperatorConfigName, Namespace: initialData.Namespace}, ocsOperatorConfig)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// If it does not exist, create the ConfigMap
+			return r.Client.Create(context.TODO(), newOCSOperatorConfig(initialData.Namespace))
+		}
+		return err
+	}
+	resourceVersion := ocsOperatorConfig.ResourceVersion
+
+	// Keep the clustername in sync with the cluster ID
+	err = r.updateClusterName(ocsOperatorConfig)
+	if err != nil {
+		return err
+	}
+	// restart the rook-ceph-operator pod if any change has occured to the configmap
+	err = r.Client.Get(context.TODO(), types.NamespacedName{Name: ocsOperatorConfigName, Namespace: initialData.Namespace}, ocsOperatorConfig)
+	if err != nil {
+		return err
+	}
+	if resourceVersion != ocsOperatorConfig.ResourceVersion {
+		return r.restartRookCephOperatorPod(initialData.Namespace)
+	}
+	return nil
+}
+
+func (r *OCSInitializationReconciler) updateClusterName(ocsOperatorConfig *corev1.ConfigMap) error {
+	clusterID := r.getClusterID()
+	// if the current value of 'CSI_CLUSTER_NAME' flag is same, just return
+	if ocsOperatorConfig.Data["CSI_CLUSTER_NAME"] == clusterID {
+		return nil
+	}
+	ocsOperatorConfig.Data["CSI_CLUSTER_NAME"] = clusterID
+	err := r.Client.Update(context.TODO(), ocsOperatorConfig)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// returns a ConfigMap for OCS operator
+// This ConfigMap is purely for the ocs-operator use only
+// No customer changes should be applied here
+func newOCSOperatorConfig(namespace string) *corev1.ConfigMap {
+	config := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      ocsOperatorConfigName,
+			Namespace: namespace,
+		},
+	}
+	data := make(map[string]string)
+	data["CSI_CLUSTER_NAME"] = ""
+	config.Data = data
+
+	return config
+}
+
+// getClusterID returns the cluster ID of the OCP-Cluster
+func (r *OCSInitializationReconciler) getClusterID() string {
+	clusterVersion := &configv1.ClusterVersion{}
+	err := r.Client.Get(context.Background(), types.NamespacedName{Name: "version"}, clusterVersion)
+	if err != nil {
+		r.Log.Error(err, "Failed to get the clusterVersion version of the OCP cluster")
+		return ""
+	}
+	return fmt.Sprint(clusterVersion.Spec.ClusterID)
+}
+
+// RestartPod deletes a pod marked by a given label
+// It then waits for a given timeout for the pod to resume running state
+func (r *OCSInitializationReconciler) restartRookCephOperatorPod(namespace string) error {
+	pod, err := r.getPodByLabel("app", "rook-ceph-operator", namespace)
```

Comment:
This whole function is unnecessary, there's no reason to try and make this a generic function when it's only used twice to get the same thing. In particular, you're already making some assumptions in that you're expecting only one Pod in the list, which I don't think you need to worry about at all. In general, I only see the following things that need to be done:

```
if configMapUpdated {
  List(podList, opts)
  Delete(podList.Items[0])

  PollImmediate(stuff, func() {
    Get(rookCephOperatorDeployment)
    if rookCephOperatorDeployment.Status.UnavailableReplicas != 0 {
      Log.Error("This could be a problem...")
      // Don't do `return err`, we can just advise it's not ready and continue reconciling other stuff.
    }
  })
}

return nil
```

In particular for 4, there's no need to return an error, we just want to mention in the logs that this is probably a bad thing, but still continue reconciliation .

Reply:
Done